docs/proposals/koordlet/20221018-koordlet-support-cgroups-v2.md

```diff
@@ -0,0 +1,461 @@
+    //  2. from lower to upper, update the new cgroup value
+    // Compatibility: if mergeFunc is nil, skip the merge and do the general update
+	LeveledUpdateBatch(cachable bool, resources [][]ResourceUpdater)
+	Run(stopCh <-chan struct{})
+}
+```
+
+Implementation of ResourceUpdateExecutor:
+
+```go
+var singleton ResourceUpdateExecutor
+
+type UpdateRequest struct {
+	Updater   ResourceUpdater
+	Cacheable bool
+}
+
+type ResourceUpdateExecutorImpl struct {
+	lock               sync.Mutex // used for leveled update
+	resourceCache      *cache.Cache
+	forceUpdateSeconds int
+	workers            int
+	updates            chan *UpdateRequest
+}
+
+func NewResourceUpdateExecutor() ResourceUpdateExecutor {
+	if singleton == nil {
+		singleton = newResourceUpdateExecutor()
+	}
+	return singleton
+}
+
+func (e *ResourceUpdateExecutorImpl) Update(cacheable bool, resource ResourceUpdater) (updated bool, err error) {
+	if cacheable {
+		return e.updateByCache(resource)
+	}
+	return true, e.update(resource)
+}
+
+func (e *ResourceUpdateExecutorImpl) Run(stopCh <-chan struct{}) {
+	go e.resourceCache.Run(stopCh)
+	for i := 0; i < e.workers; i++ {
+		go e.process(stopCh)
+	}
+	<-stopCh
+	close(e.updates)
+}
+```
+
+#### Cgroup Read
+
+**CgroupReader** defines unified cgroup reads operations.
+
+```go
+type CgroupReader interface {
+	ReadCPUQuota(parentDir string) (int64, error)
+	ReadCPUAcctStat(parentDir string) (*CPUAcctStat, error)
+	ReadCPUStat(parentDir string) (*CPUStat, error)
+	ReadMemoryLimit(parentDir string) (int64, error)
+	ReadMemoryStat(parentDir string) (*MemoryStat, error)
+}
+```
+
+CgroupReader Implementations:
+
+```go
+func NewCgroupReader() CgroupReader {
+	if sysutil.Conf.UseCgroupV2 {
+		return &CgroupV2Reader{}
+	}
+	return &CgroupV1Reader{}
+}
+
+// for cgroups-v1
+type CgroupV1Reader struct{}
+
+func (r *CgroupV1Reader) ReadCPUQuota(parentDir string) (int64, error) {
+	s, err := sysutil.CgroupFileRead(parentDir, systemresource.CPUCFSQuota)
+	if err != nil {
+		return -1, fmt.Errorf("cannot read cgroup file, err: %v", err)
+	}
+	// content: "-1", "100000", ...
+	v, err := strconv.ParseInt(s, 10, 64)
+	if err != nil {
+		return -1, fmt.Errorf("cannot parse cgroup value %s, err: %v", s, err)
+	}
+	return v, nil
+}
+
+// for cgroups-v2
+type CgroupV2Reader struct{}
+
+func (r *CgroupV2Reader) ReadCPUQuota(parentDir string) (int64, error) {
+	s, err := sysutil.CgroupFileRead(parentDir, systemresource.CPUMaxV2)
+	if err != nil {
+		return -1, fmt.Errorf("cannot read cgroup file, err: %v", err)
+	}
+	// content: "max 100000", "100000 100000"
+	ss := strings.Fields(s)
+	if len(ss) != 2 {
+		return -1, fmt.Errorf("cannot parse cgroup value %s, err: invalid pattern", s)
+	}
+	if ss[0] == sysutil.CgroupMaxSymbolStr {
+		return -1, nil
+	}
+	v, err := strconv.ParseInt(ss[0], 10, 64)
+	if err != nil {
+		return -1, fmt.Errorf("cannot parse cgroup value %s, err: %v", s, err)
+	}
+	return v, nil
+}
+```
+
```

Comment:
how about add a using case (in qos plugin) of the API

Reply:
done